src/main/kotlin/org/jetbrains/kotlinx/jupyter/execution.kt

```diff
@@ -0,0 +1,72 @@
+package org.jetbrains.kotlinx.jupyter
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.jetbrains.kotlinx.jupyter.exceptions.ReplException
+import java.util.Collections
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.concurrent.thread
+
+sealed interface ExecutionResult<out T> {
+    class Success<out T>(val result: T) : ExecutionResult<T>
+    class Failure(val throwable: Throwable) : ExecutionResult<Nothing>
+    object Interrupted : ExecutionResult<Nothing>
+}
+
+interface JupyterExecutor {
+    fun <T> runExecution(classLoader: ClassLoader? = null, body: () -> T): ExecutionResult<T>
+    fun interruptExecutions()
+
+    fun launchJob(runnable: suspend CoroutineScope.() -> Unit)
+}
+
+class JupyterExecutorImpl : JupyterExecutor {
+    private val currentExecutions: MutableSet<Thread> = Collections.newSetFromMap(ConcurrentHashMap())
+    private val coroutineScope = CoroutineScope(Dispatchers.Default)
+
+    override fun <T> runExecution(classLoader: ClassLoader?, body: () -> T): ExecutionResult<T> {
+        var execRes: T? = null
+        var execException: Throwable? = null
+        val execThread = thread(contextClassLoader = classLoader ?: Thread.currentThread().contextClassLoader) {
```

Comment:
Sorry for posting a bit late. This would work, but creating and stopping a new thread a bit costly since it's OS call. 

The more consise would be to reuse existing threads, so they wait for some tasks to appear from a queue and execute them, e.g. pool aproach. So perhaps we might reuse some existing, like `Executors.newCachedThreadPool `

Reply:
Yes, I undestand. It's just how it was implemented before. I will change it in a separate commit